### PR TITLE
Add stop_on_first_failure optn to test_descriptor

### PR DESCRIPTION
**Pull request by kunaljdoshi**
If stop_on_first_failure=True, the test executor exits
the test on first failed measurement phase. This is useful
during debugging

---
This change is [Reviewable](https://reviewable.io/reviews/google/openhtf/879)

### CODE REVIEW

File: openhtf/core/test_executor.py [hunk 1]
```diff
@@ -233,6 +235,14 @@ def _handle_phase(self, phase):
 
     self.test_state.state_logger.debug('Handling phase %s', phase.name)
     outcome = self._phase_exec.execute_phase(phase)
+    if self.test_state.test_options.stop_on_first_failure:
+      # Stop Test on first measurement failure
+      current_phase_result = self.test_state.test_record.phases[
+          len(self.test_state.test_record.phases) -1]
+      if current_phase_result.outcome == openhtf.test_record.PhaseOutcome.FAIL:
+        outcome = phase_executor.PhaseExecutionOutcome(openhtf.PhaseResult.STOP)
+        self.test_state.state_logger.debug('Stopping test due to failed measurement')
```
Comment (arsharma1) [inline openhtf/core/test_executor.py]: Please limit lines to 80 characters.

Comment (arsharma1) [inline openhtf/core/test_executor.py]: Please log at  `error` instead of `debug`.

File: openhtf/core/test_executor.py [hunk 2]
```diff
@@ -18,8 +18,10 @@
 import sys
 import threading
 
+import openhtf
```
Comment (arsharma1) [inline openhtf/core/test_executor.py]: Don't import the base openhtf module; we are trying to remove the circular imports.

File: openhtf/core/test_executor.py [hunk 3]
```diff
@@ -233,6 +235,14 @@ def _handle_phase(self, phase):
 
     self.test_state.state_logger.debug('Handling phase %s', phase.name)
     outcome = self._phase_exec.execute_phase(phase)
+    if self.test_state.test_options.stop_on_first_failure:
+      # Stop Test on first measurement failure
+      current_phase_result = self.test_state.test_record.phases[
+          len(self.test_state.test_record.phases) -1]
+      if current_phase_result.outcome == openhtf.test_record.PhaseOutcome.FAIL:
```
Comment (arsharma1) [inline openhtf/core/test_executor.py]: test_record is already imported.

... == test_record.PhaseOutcome.FAIL

File: openhtf/core/test_executor.py [hunk 4]
```diff
@@ -233,6 +235,14 @@ def _handle_phase(self, phase):
...
+        outcome = phase_executor.PhaseExecutionOutcome(openhtf.PhaseResult.STOP)
```
Comment (arsharma1) [inline openhtf/core/test_executor.py]: openhtf.PhaseResult.STOP  -> phase_descriptor.PhaseResult.STOP

phase_descriptor will need to be imported

File: openhtf/core/test_executor.py (hunk 1 above)
Comment (kunaljdoshi) [reply openhtf/core/test_executor.py]: Done

File: openhtf/core/test_executor.py (hunk 2 above)
Comment (kunaljdoshi) [reply openhtf/core/test_executor.py]: Done

File: openhtf/core/test_executor.py (hunk 4 above)
Comment (kunaljdoshi) [reply openhtf/core/test_executor.py]: Done

File: examples/stop_on_failure.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+from openhtf.plugs import user_input
+from openhtf.util import validators
+from openhtf.output.callbacks import console_summary
+
+# One can use phase option stop_on_failure
+# to stop test execution on failed measurement
+@htf.PhaseOptions(stop_on_failure=True)
+@htf.measures('number_sum', validators=[validators.in_range(0, 5)])
+def add_numbers(test):
```
Comment (arsharma1) [inline examples/stop_on_failure.py]: phase_that_fails

Make it obvious that this is expected to fail.

File: openhtf/core/test_executor.py [hunk 6]
```diff
@@ -233,6 +239,17 @@ def _handle_phase(self, phase):
 
     self.test_state.state_logger.debug('Handling phase %s', phase.name)
     outcome = self._phase_exec.execute_phase(phase)
+    if (self.test_state.test_options.stop_on_first_failure or
+        conf.stop_on_first_failure):
+      # Stop Test on first measurement failure
+      current_phase_result = self.test_state.test_record.phases[
+          len(self.test_state.test_record.phases) -1]
```
[the remark this answers is not in the document]
Comment (kunaljdoshi) [reply openhtf/core/test_executor.py]: Done. PTAL.

**Pull request merged**